---
  title: "Workflow"
author: "Basil Khuder"
date: "7/29/2020"
output:
  html_document:
  toc: true
---
  
This is the workflow used for the BAL single-cell analysis. Please install the following packages in order to recreate this analysis.
  
```{}
install.packages("devtools")
install.packages("Seurat")
install.packages("tidyverse")
install.packages("Matrix")
install.packages("data.table")
install.packages("pheatmap")
devtools::install_github("jokergoo/ComplexHeatmap")
install.packages("Cairo")
install.packages("cowplot)
BiocManager::install("org.Hs.eg.db")
BiocManager::install("org.Hs.eg.db")
BiocManager::install("clusterProfiler")
BiocManager::install("SingleR")
```


# Importing the Data and loading the libraries 
The data is uploaded to GEO and should be moved to the working directory where you will be running this
analysis. Please create a new folder within the working directory named as "GEO_data", and put the data from GEO in this folder.

We must also import all of the required packages:

```{r}
library(Seurat)
library(pheatmap)
library(ComplexHeatmap)
library(Matrix)
library(tidyverse)
library(data.table)
library(SingleR)
library(EnhancedVolcano)
library(Cairo)
library(cowplot)
library('org.Hs.eg.db')

source('annotation.R')
source('data_import.R')
source('extraction.R')
source('processing.R')
source('utilities.R')
source('visualization.R')

```

# Creating the Seurat object
To create our Seurat object, we must load in our data. Using crToSeurat's ```merge```
parameter will also merge all of our objects together.

```{r}
bal.all.control <- crToSeurat(directory = "./GEO_data",
                    sample_names = c("BAL01", "BAL02","BAL03","BAL04"),
                    merge = TRUE)
```

# Filtration 

Our filtration method uses quantile percentages for each individual sample, rather than
a set arbitrary number (as done in the Seurat vignette.) After playing around with
the parameters, we choose to filter cells with levels of mitochondrial genes in the
bottom 5% and top 97.5% for each sample. We also filtered cells with features in 
the top 97.5%

```{r cache=TRUE}
bal.all.control <- featureFiltration(bal.all.control,
                           mito.low = 0.05, 
                           mito.high = .975, 
                           feature.cut = 0.975,
                           produce_plots = TRUE) 
```

# Downstream Normalization and Processing 

```processSeurat()``` uses SCTransform for normalization (which allows us to be less
conservative with choosing principal components.) .2 cluster resolution gave us
the most representative cell clustering. If you want to play around with other
cluster resolutions, use the chooseClusterResolution(). 

```{r cache=TRUE, fig.width=7, fig.height=7}
bal.all.control <- processSeurat(bal.all.control,
                       dim = 1:30, 
                       cluster.res = .2, 
                       vars_to_regress = c("orig.ident","percent.mt"))
DimPlot(bal.all.control)
FeaturePlot(bal.all.control, features = c("FABP4","CCL3","CCL4","LTB","MARCKS","MS4A1","SLPI","SCGB3A1","CAPS"), ncol = 3)
```

# Cluster Markers
To see what highly expressed genes for each cluster, we run the ```produceMarkers()```
function. This will also save a file called 'markers.txt'

```{r}
produceMarkers(bal.all.control, 
               cells.per.ident = Inf, 
               top.gene.plot = TRUE,
               file.name = "markers.txt")
```

With the representive markers for each cluster, we can rename our clusters for the cell-type: 

```{r}
bal.all.control <- renameClusters(bal.all.control,
                                  cluster_names  = c("Macrophages", "Macrophages", "T/NK Cells",
                                 "Macrophages", "Macrophages", "Macrophages",
                                 "Macrophages", "Dendritic Cells", "Macrophages",
                                 "Dividing Cells", "Club Cells", "B Cells", 
                                 "Ciliated Cells"))
```

```{r}
DimPlot(bal.all.control, label = TRUE)
```

We're interested in an inflammatory subset of Macrophages called MIP so we'll subset our Macrophages cluster to try to identify them. Through trial, we chose
a cluster resolution of .03. 

```{r}
bal.mp <- subset(bal.all.control, idents = "Macrophages")
bal.mp <- processSeurat(bal.mp,
                       dim = 1:30, 
                       cluster.res = .03, 
                       vars_to_regress = c("orig.ident","percent.mt"))
produceMarkers(bal.mp, 
               cells.per.ident = Inf, 
               top.gene.plot = TRUE,
               file.name = "MIP_markers.txt")
DimPlot(bal.mp, label = TRUE)
```

We can also use Seurat's "Module Scores" function to see the expression of a list
of genes. 

```{r}
plotModuleScores(bal.mp, 
                 features = c("CCL4","CCL3","CXCL10","CCL20","CCL4L2"))
```

We can now rename the MIP clusters, and merge the MIP identities back into the 
original bal.all.control object. 

```{r}
bal.mp <- renameClusters(bal.mp, cluster.names = c("Macrophages","MIP-1 Macrophages"))
bal.all.control <- mergeSubCluster(bal.all.control, 
                                   bal.mp,
                                   "Macrophages")
DimPlot(bal.all.control, label = TRUE)
```

# Figure Recreation

All code that was used to create figures can be found in the "figures.R" and "figures_png.R" files.
The only figure that was made outside of R was the GSEA plots. To recreate those figures, run the code below to generate a pre-ranked list of differential genes in the Macrophage cluster, and import this file into GSEA.


## GSEA Pre-ranked list (figure made in the GSEA program, not R.)
```{r}
mip.all.markers <- FindMarkers(bal.mp, ident.1 = "MIP-1 Macrophages", ident.2 = "Macrophages")
mip1.gsea.rnk <- data.frame(row.names(mip.all.markers), mip.all.markers$avg_logFC)
write.table(mip1.gsea.rnk, file="mip_gsea.rnk", row.names=FALSE, col.names = FALSE, sep="\t")
```




## SingleR HPCA Figures

```{r}
bal.sce <- as.SingleCellExperiment(bal.all.control)
hpca.se <- HumanPrimaryCellAtlasData()
pred.bal.all <- SingleR(test = bal..sce, ref = hpca.se, labels = hpca.se$label.main,
                        method = "single")

heatmap.colors = colorRamp2(seq(-1,2,length.out =100), colorRampPalette(rev(brewer.pal(n = 7, name ="RdYlBu")))(100))
Heatmap(pred.bal.all$scores, col = heatmap.colors, heatmap_legend_param = list(title = ""), show_row_names = FALSE,  column_names_side = "top", column_dend_side = "bottom", column_names_rot = 45)
```






## MIP Macrophage Differential Gene Volcano Plot
```{r}
mip.all.markers <- FindMarkers(bal.all.control, ident.1 = "MIP Macrophages", ident.2 = "Macrophages",
                                )

mip.all.markers <- cbind(gene = row.names(mip1.all.markers), mip1.all.markers) 
EnhancedVolcano(mip.all.markers,
                lab = rownames(mip.all.markers),
                x = 'avg_logFC',
                y = 'p_val_adj',
                xlim = c(-1, 1))

EnhancedVolcano(mip.all_markers,
                lab = rownames(mip.all_markers),
                x = 'avg_logFC',
                y = 'p_val_adj',
                title = "Differential Expression Between MIP and Macrophages", subtitle = '',
                FCcutoff = 1,
                xlim = c(-1, 2.2),
                pointSize = 3.0,
                selectLab = c("MCEMP1","TXNIP","CCL4","CCL3","CXCL10"),
                xlab = bquote('Avg' ~Log[2]~ 'Fold Change'),
                ylab = bquote(~-Log[10]~Adjusted~italic(P-Value)),
                legendLabels = c("NS","Adj P-Value","Adj P-Value",'Adj P-Value & Avg Log (base 2) Fold-Change'),
                legendPosition = 'top',
                legendLabSize = 14,
                labSize = 3.0)
```